compiler/p/codegen/PPCBinaryEncoding.cpp

```diff
@@ -619,229 +945,410 @@ TR::PPCTrg1Src1ImmInstruction::addMetaDataForCodeAddress(uint8_t *cursor)
+   mask &= 0xffffffff;
+   TR_ASSERT_FATAL_WITH_INSTRUCTION(instr, mask != 0, "Cannot encode a mask of 0");
+
+   uint32_t mask32 = static_cast<uint32_t>(mask);
+
+   if (mask32 != 0xffffffffu && (mask32 & 0x80000001u) == 0x80000001u)
       {
-      setSourceImmediate(getSourceImmediate() << 10);
-      insertImmediateField(toPPCCursor(cursor));
+      int32_t lead = leadingZeroes(~mask32);
+      int32_t trail = trailingZeroes(~mask32);
+
+      TR_ASSERT_FATAL_WITH_INSTRUCTION(instr, mask32 == ~((0xffffffffu >> lead) & (0xffffffffu << trail)), "Mask of 0x%x has more than one group of 1 bits", mask32);
+
+      return std::make_pair(32 - trail, lead - 1);
       }
    else
       {
-      insertImmediateField(toPPCCursor(cursor));
-      }
+      int32_t lead = leadingZeroes(mask32);
+      int32_t trail = trailingZeroes(mask32);
 
-   addMetaDataForCodeAddress(cursor);
+      TR_ASSERT_FATAL_WITH_INSTRUCTION(instr, mask32 == ((0xffffffffu >> lead) & (0xffffffffu << trail)), "Mask of 0x%x has more than one group of 1 bits", mask32);
 
-   cursor += PPC_INSTRUCTION_LENGTH;
-   setBinaryLength(PPC_INSTRUCTION_LENGTH);
-   setBinaryEncoding(instructionStart);
-   return cursor;
+      return std::make_pair(lead, 31 - trail);
+      }
    }
 
-
-static void insertMaskField(uint32_t *instruction, TR::InstOpCode::Mnemonic op, int64_t lmask)
+void TR::PPCTrg1Src1Imm2Instruction::fillBinaryEncodingFields(uint32_t *cursor)
    {
-   int32_t encoding;
-   // A mask is is a string of 1 bits surrounded by a string of 0 bits.
-   // For word instructions it is specified through its start and stop bit
-   // numbers.  Note - the mask is considered circular so the start bit
-   // number may be greater than the stop bit number.
-   // Examples:     input     start   stop
-   //              00FFFF00      8     23
-   //              00000001     31     31
-   //              80000001     31      0
-   //              FFFFFFFF      0     31  (somewhat arbitrary)
-   //              00000000      ?      ?  (illegal)
-   //
-   // For doubleword instructions only one of the start bit or stop bit is
-   // specified and the other is implicit in the instruction.  The bit
-   // number is strangely encoded in that the low order bit 5 comes first
-   // and the high order bits after.  The field is in bit positions 21-26.
+   TR::RealRegister *trg = toRealRegister(getTargetRegister());
+   TR::RealRegister *src = toRealRegister(getSource1Register());
+   uint32_t imm1 = getSourceImmediate();
+   uint64_t imm2 = getLongMask();
 
-   // For these instructions the immediate is not a mask but a 1-bit immediate operand
-   if (op == TR::InstOpCode::cmprb)
+   switch (getOpCode().getFormat())
       {
-      // populate 1-bit L field
-      encoding = (((uint32_t)lmask) & 0x1) << 21;
-      *instruction |= encoding;
-      return;
-      }
+      case FORMAT_RLDIC:
+         {
```

Comment:
these kinds of one FORMAT per instruction are too tedious and somewhat misleading anyway.  MD-form and MDS-form are good enough to distinguish all of them.  Although MDS-form includes more instructions beyond the rotate/shift, but we have never used them in practice.  Thus, MDS-form can be used to signify them exclusively.

Reply:
As much as I'd love to agree (and it was actually something I tried myself while working on this), the way we currently handle immediates on rotate-and-mask instructions makes this impossible. If we specified immediates in the same manner as the ISA, then I'd definitely agree that these should be removed. Instead, we specify immediates by passing the mask to be used as a parameter, like so:

```
rldicr gr0, gr0, 8, 0xffffffff80000000
rldicl gr0, gr0, 8, 0x00000000ffffffff
rldic  gr0, gr0, 8, 0x00000000ffffff00
```

All of these instructions use the same value for the `me`/`mb` field (32), however they all have wildly different immediate values in their `TR::Instruction` objects. Since the immediate specifies a mask rather than the bit number that the ISA specifies should be encoded in the instruction, we need to know how to derive that bit number from the mask. Because of that, these instructions simply can't be handled in a uniform manner given the way we currently generate them.

Effectively, this is just a path forward to maintain compatibility with our old code, which had special cases for these particular instructions. The reason behind using a format for these rather than using special cases that check the instruction opcode directly (beyond just trying to keep special cases to a minimum) is to reduce the room for error and complexity with regard to the record-form variants of these opcodes. Since record-form opcodes inherit the format of their non-record-form counterparts, they work without needing any special consideration. Failing to check for the record-form variant of an opcode when special casing turned out to be a surprisingly common mistake in the old binary encoder.

Reply:
it can be solved seamlessly to split MDS-form into two formats:  MDS_MB  and MDS_ME (or something like that). This way, the mask immediate can be verified too (asserted), knowing whether you are looking for MB or ME.

Reply:
Using two formats based on whether to use `me` or `mb` still isn't enough, as it does not cover `rldic`, which uses a mask begin and has an _implicit_ mask end based on the shift amount that needs to be validated. Unless a separate format is used for `rldic`, it is simply not possible to verify that the mask is correct, which leaves us with the same number of formats we started with. Renaming these also wouldn't help IMO as the encoding of these formats is still implicitly tied to the behaviour of the instruction in question, since that's what determines what mask is really used.

I agree that this is not an ideal solution. I've already wasted countless days going down this path and trying to correct this to the best of my ability, but I was unable to find a better solution that could be implemented without significant time investment. As far as I can tell, the only options are to either use separate formats for each or change how we interpret the immediates on these instructions.

Reply:
as a compromise, we can work toward splitting MDS form, but verifying the mask according to opcode in MDS_MB.

Reply:
I don't think that compromise addresses my concerns. How the value of the `mb`/`me` field is derived from the bitmask is still intrinsically an implementation detail of each individual instruction. Naming the formats as if they were general gives the false impression that they are. Reducing this to two formats would also reintroduce the need for more special cases based on checking the opcode, something I've been trying to keep to an absolute minimum for reasons I've already laid out.

As I've already explained, what I'm currently doing is not a good solution, but that's mostly down to how we use the immediates. These formats are _not general_ and are warts that are only being preserved right now so as to keep compatibility and to not expand the scope of this PR any more than it already has. Renaming these formats to appear more general than they really are would only mask the underlying problem. The names of these formats are not clean because the formats themselves are not clean.

If it helps, I can open an issue with a concrete plan to have these instructions migrated to match the ISA and then have the formats renamed at that time.

Reply:
ok.  as a temporary solution, fine.  but, at end of the day, you cannot argue from both ends (general-purpose, and specific-instruction hindering the ISA future perspective). plus, it looks to me that per-opcode assertion is not grossly bad.

Reply:
I've opened #4976 with a plan to migrate to match the ISA and merge/rename these formats.